src/deepsparse/server/server.py

```diff
@@ -0,0 +1,159 @@
+        "handlers": ["default"],
+        "level": log_level.upper(),
+    }
+
+    with open(config_path) as fp:
+        obj = yaml.safe_load(fp)
+    server_config = ServerConfig(**obj)
+    _LOGGER.info(f"Using config: {repr(server_config)}")
+
+    app = _build_app(server_config)
+
+    uvicorn.run(
+        app,
+        host=host,
+        port=port,
+        log_level=log_level,
+        log_config=log_config,
+        # NOTE: only want to have 1 server process so models aren't being copied
+        workers=1,
+    )
+
+
+def _build_app(server_config: ServerConfig) -> FastAPI:
+    context = Context(
+        num_cores=server_config.num_cores,
+        num_streams=server_config.num_workers,
+        scheduler=Scheduler.multi_stream,
+    )
+
+    _LOGGER.info(f"Built context: {repr(context)}")
+
+    app = FastAPI()
+
+    @app.get("/", include_in_schema=False)
+    def _home():
+        return RedirectResponse("/docs")
+
+    @app.get("/config", tags=["general"], response_model=ServerConfig)
+    def _info():
+        return server_config
+
+    @app.get("/ping", tags=["general"], response_model=bool)
+    @app.get("/health", tags=["general"], response_model=bool)
+    @app.get("/healthcheck", tags=["general"], response_model=bool)
+    @app.get("/status", tags=["general"], response_model=bool)
+    def _health():
+        return True
+
+    app._deepsparse_pool = ThreadPoolExecutor(max_workers=context.num_streams)
+
+    add_invocations_endpoint = len(server_config.endpoints) == 1
+
+    # fill in names if there are none
+    for idx, endpoint_config in enumerate(server_config.endpoints):
+        if endpoint_config.name is None:
+            endpoint_config.name = f"endpoint-{idx}"
+
+    # creat pipelines & endpoints
+    for endpoint_config in server_config.endpoints:
+        pipeline_config = endpoint_config.to_pipeline_config()
+        pipeline_config.kwargs["executor"] = app._deepsparse_pool
+
+        _LOGGER.info(f"Initializing pipeline for '{endpoint_config.name}'")
+        pipeline = Pipeline.from_config(pipeline_config, context)
+
+        _LOGGER.info(f"Adding endpoints for '{endpoint_config.name}'")
+        _add_pipeline_endpoint(app, endpoint_config, pipeline, add_invocations_endpoint)
+
+    _LOGGER.info(f"Added endpoints: {[route.path for route in app.routes]}")
+
+    return app
+
+
+def _add_pipeline_endpoint(
+    app: FastAPI,
+    endpoint_config: EndpointConfig,
+    pipeline: Pipeline,
+    add_invocations_endpoint: bool = False,
+):
+    input_schema = pipeline.input_schema
+    output_schema = pipeline.output_schema
+
+    pool: ThreadPoolExecutor = app._deepsparse_pool
+
+    route = endpoint_config.endpoint or "/predict"
+    if not route.startswith("/"):
+        route = "/" + route
+
+    @app.post(route, tags=["predict"], response_model=output_schema)
+    async def _predict_func(request: pipeline.input_schema):
+        return await get_running_loop().run_in_executor(pool, pipeline, request)
```

Comment:
will using the same threadpool for the run loop as well as inside the pipeline (ie for dyamic) give us an inherent "minus 1" on concurrent inferences since one worker will be dedicated to this main predict func?

Reply:
Hmm good question. Unclear how this all works with multiple concurrent server requests. We may also just be able to remove this async run_in_executor thing and do `pipeline(request)`. I think this needs some benchmarking to say.

Reply:
I just ran some benchmarking, and this version performed the same as just calling `pipeline(request)`, so I replaced it to not use the async version.